tsl/src/nodes/decompress_chunk/decompress_chunk.c

```diff
@@ -359,6 +359,84 @@ cost_decompress_chunk(Path *path, Path *compressed_path)
+/*
+ * When limit queries use 'order by' and the first attribute of the order by is equivalent to the
+ * first attribute of the hypertable it can be beneficial to sort chunks individually. In that case
+ * ChunkAppend can be used and only the needed chunks are accessed (e.g., if the query has a small
+ * limit value, only the first chunk needs to be accessed in most cases).
+ *
+ * If at least one chunk of a hypertable is uncompressed, PostgreSQL will generate a MergeAppend
+ * path in generate_orderedappend_paths() / create_merge_append_path() due to the existing
+ * pathkeys of the index of the uncompressed chunk. If possible, the merge append operation is
+ * performed by the ChunkAppend node, and only the needed chunks are accessed.
+ *
+ * If all chunks are compressed, none of the chunks has pathkeys and therefore no MergeAppend
+ * path is generated by PostgreSQL. As a result, an expensive (parallel) query plan with the
+ * sort node on top of the DecompressChunkNodes and below the limit node is chosen, which
+ * decompresses all the chunks.
+ *
+ * This function adds a sorted path to the chunk if this is beneficial. As a result, this chunk has
+ * path keys and PostgreSQL will generate a MergeAppendPath that can be performed by ChunkAppend.
+ */
+static void
+add_chunk_sorted_path(PlannerInfo *root, RelOptInfo *chunk_rel, Hypertable *ht, Chunk *chunk,
+					  Index ht_relid, Path *decompress_chunk_path)
+{
+	if (root->limit_tuples <= 0 || root->query_pathkeys == NIL || ts_chunk_is_partial(chunk) ||
+		hypertable_is_distributed(ht))
+		return;
+
+	ListCell *lc;
+
+	SortGroupClause *sort = linitial(root->parse->sortClause);
+	TargetEntry *tle = get_sortgroupref_tle(sort->tleSortGroupRef, root->parse->targetList);
+
+	if (!IsA(tle->expr, Var))
+		return;
+
+	Var *sort_var = castNode(Var, tle->expr);
+	Index sort_relid = sort_var->varno;
+
+	/* If sort happens on another table, we could not optimize the case */
+	if (sort_relid != ht_relid)
+		return;
+
+	if (!AttrNumberIsForUserDefinedAttr(sort_var->varattno))
+		return;
+
+	/* Check hypertable column is the first dimension of the hypertable */
+	RangeTblEntry *rte = root->simple_rte_array[ht_relid];
+	char *column =
+		strVal(list_nth(rte->eref->colnames, AttrNumberGetAttrOffset(sort_var->varattno)));
+
+	if (namestrcmp(&ht->space->dimensions[0].fd.column_name, column) != 0)
+		return;
+
+	/* Check if we can perform a sort on the query pathkeys */
+	foreach (lc, root->query_pathkeys)
+	{
+		PathKey *pathkey = (PathKey *) lfirst(lc);
+		EquivalenceClass *pathkey_ec = pathkey->pk_eclass;
+
+		if (pathkey_ec->ec_has_volatile)
+			return;
+
+		Expr *em_expr = find_em_expr_for_rel(pathkey_ec, chunk_rel);
+
+		if (!em_expr)
+			return;
+	}
+
+	/* Create the sorted path */
+	DecompressChunkPath *dcpath =
+		copy_decompress_chunk_path((DecompressChunkPath *) decompress_chunk_path);
+
+	Path *sorted_path =
+		(Path *) create_sort_path(root, chunk_rel, &dcpath->cpath.path, root->query_pathkeys, -1);
```

Comment:
I think it would worth the effort to pass `root->limit_tuples` in case its possible.
...as switching to a `topN` instead of a full sort is huge difference! 

iirc I've seen somewhere some caveats regarding this - but don't know where...